### ISSUE TITLE
RTD functions don't work when Excel requires signing

### ISSUE THREAD

**Issue by evolvedlight** [Excel-DNA/ExcelDna#298]
Hi

This is kinda similar to #141 but I thought I'd writeup exactly why this isn't working:

Sample testcase:

- Start Excel, ensure in trustcenter that "Require Application Add.ins to be signed by Trusted Publisher" is ticked.
- Create new ExcelDna addin as per normal
Add code:
```
        [ExcelFunction(Description = "My first .NET function")]
        public static object SayHelloAsync(string name)
        {
            return ExcelAsyncUtil.Run("RunSomethingDelay", null, () => RunSomethingDelay());
        }

        public static string RunSomethingDelay()
        {
            Thread.Sleep(2000);
            return "test";
        }
```
Try and use it: you should get the following warning:
![image](https://user-images.githubusercontent.com/23338/74539186-e3b34f00-4f3d-11ea-9129-438bb10e64c6.png)

We see that the call succeeded:
![image](https://user-images.githubusercontent.com/23338/74539163-ddbd6e00-4f3d-11ea-92cd-27ef7a4105be.png)

We then don't actually manage to create the RTD server, but Excel reports success. We add this to loadedRtdServers, so when further calls try and use the RTD functions (after accepting the security warning), we try and use the RTD server that never started.

I'd very much like to try and fix this, any guidance to the preferred way? It seems that somehow we need to track the point of the RTD server actually starting up, not just the request to start it up. Otherwise I'll see if I can create something to fix this.

Thanks,
Steve

**Comment by govert** [comment]
The "XlReturnSuccess" merely means that the C API call did not fail.
For an RTD server that could not be loaded (e.g. a ProgId that does not exists) the TryExcel will return Success, but "result" will be `ExcelError.ExcelErrorNA`.

It would be possible to have some global (add-in wide) flag that indicates whether the internal RTD server ( `ExcelObserverRtdServer`) has started or not. Then one might have some mechanism to display the error to the user somehow.

**Comment by evolvedlight** [comment]
Hi @govert 
Thanks - I see that, thanks. I didn't notice that `result`. 
I'll try a better pull request..
Thanks,
Steve

**Comment by govert** [comment]
Note that there are be legitimate cases for the result to be `ExcelErrorNA` if the server started correctly too.

**Comment by evolvedlight** [comment]
@govert so the only way to actually know if the RTD server was started is to register it at a similar point as in my PR? If we can't rely on it not being ExcelErrorNA then I don't see any better solution.

**Comment by govert** [comment]
Yes, something like that. How do you want to use the fact that it failed to start?

**Comment by evolvedlight** [comment]
@govert I don't want to use the fact in my own code: I want the library to realise that the RTD server didn't start, and thus the next time try again to start an RTD server. Inbetween those two requests I expect the user to click the "Enable" button to allow RTD to work.

Ideally we'd know that it didn't work because of the permissions problem, and display that back to the user to click accept and try again, but that's secondary to this working at all. At the moment, if corporate policy doesn't allow turning off the signed addins thing, you can't use the Async tasks support at all, even if the addin is signed (which seems weird)

**Comment by govert** [comment]
Signed add-ins should work if the Trust Center settings are right, IIRC. The settings for normal add-ins and the COM parts like RTD are different.

Anyway, so it would work if you had some flag exposed on `ExcelAsyncUtil`, e.g. `ExcelAsyncUtil.ServerLoaded'?
Or, I guess more difficult, maybe `ExcelAsyncUtil.ServerLoadFailed`.

**Comment by evolvedlight** [comment]
@govert I'm sure you're right and there's just something wrong in the corp workplace - but for now at least ensuring that we don't try to use broken RTD servers is good enough. 

Could you check the PR #299 please to see if it looks suitable? I don't address yet any feedback mechanism.

**Comment by govert** [comment]
For the build-in `ExcelObserverRtdServer` we expect at most one instance to be loaded at a time. Maybe we can just have this one instance report whether it has started via some exported method? That would reduce the amount of fiddling in the RTD registration re-direct code, which I'd prefer.

To decide on that, I think you first need to think about what you want to know from your add-in side, and how that should be exposed.

**Comment by evolvedlight** [comment]
Hi @govert ,

Thanks for your patience here, although I'm not sure I totally understand what you suggest here.
Ignoring anything I'm doing here specifically (let's assume I have no requirements from my own addin apart from "it works, I think there's still the bug if people try and use the library to create their own RTD servers - they won't actually work the first time it's called. I think therefore that this code has to go somewhere inside `ExcelRtd`? It's exactly within the RTD registration redirect code that I think this problem is.

Regards,
Steve
